probe-rs/src/architecture/arm/sequences/atsame5x.rs

```diff
@@ -0,0 +1,379 @@
+    }
+}
+
+impl DsuStatusB {
+    const ADDRESS: u64 = 0x4100_2102;
+}
+
+/// A wrapper for different types that can perform SWD Commands (SWJ_Pins SWJ_Sequence)
+struct SwdSequenceShim<'a>(&'a mut dyn architecture::arm::communication_interface::DapProbe);
+
+impl<'a> From<&'a mut dyn architecture::arm::communication_interface::DapProbe>
+    for SwdSequenceShim<'a>
+{
+    fn from(p: &'a mut dyn architecture::arm::communication_interface::DapProbe) -> Self {
+        Self(p)
+    }
+}
+
+impl<'a> architecture::arm::communication_interface::SwdSequence for SwdSequenceShim<'a> {
+    fn swj_sequence(&mut self, bit_len: u8, bits: u64) -> Result<(), Error> {
+        self.0.swj_sequence(bit_len, bits).map_err(Error::Probe)
+    }
+
+    fn swj_pins(&mut self, pin_out: u32, pin_select: u32, pin_wait: u32) -> Result<u32, Error> {
+        self.0
+            .swj_pins(pin_out, pin_select, pin_wait)
+            .map_err(Error::Probe)
+    }
+}
+
+/// Marker struct indicating initialization sequencing for ATSAM D5x/E5x family parts.
+pub struct AtSAME5x {}
+
+impl AtSAME5x {
+    /// Create the sequencer for the ATSAM D5x/E5x family of parts.
+    pub fn create() -> Arc<Self> {
+        Arc::new(Self {})
+    }
+
+    /// Perform a Chip-Erase operation
+    ///
+    /// # Errors
+    /// This operation can fail due to insufficient permissions, or if Chip-Erase Lock is
+    /// enabled (this lock can only be released from within the device firmare).
+    /// After a successful Chip-Erase a `DebugProbeError::ReAttachRequired` is returned
+    /// to signal that a re-connect is needed for the DSU to start operating in unlocked mode.
+    pub fn erase_all(&self, memory: &mut Memory, permissions: &Permissions) -> Result<(), Error> {
+        let dsu_status_b = DsuStatusB::from(memory.read_word_8(DsuStatusB::ADDRESS)?);
+
+        match (dsu_status_b.celck(), dsu_status_b.prot(), permissions.erase_all()) {
+            (true, _, _) => Err(Error::MissingPermissions(
+                "Chip-Erase is locked. This can only be unlocked from within the device firmware by performing \
+                a Chip-Erase Unlock (CEULCK) command."
+                    .into(),
+            )),
+            (false, true, Err(_)) => Err(Error::MissingPermissions(
+                "Device is locked. A Chip-Erase operation is required to unlock. \
+                            Re-run with '--connect-under-reset --allow-erase-all' to perform Chip-Erase."
+                    .into(),
+            )),
+            (false, false, Err(e)) => Err(e),
+            (false, _, Ok(())) => Ok(()),
+        }?;
+
+        // Start the chip-erase process
+        let mut dsu_ctrl = DsuCtrl(0);
+        dsu_ctrl.set_ce(true);
+        memory.write_word_8(DsuCtrl::ADDRESS, dsu_ctrl.0)?;
+        log::info!("Chip-Erase started..");
+
+        // Wait for it to finish
+        let start = std::time::Instant::now();
+        while start.elapsed() < std::time::Duration::from_secs(8) {
```

Comment:
Yes this can probably be done better. Chip-Erase takes around 3-5 seconds on my boards (1024KB Flash). My idea was to set a reasonable upper limit, but then check relatively often to detect quickly when its done.  Can you point me to an example in the code that I can mimic instead?

Reply:
Actually while I was going through code to look for an example, I actually found that many use "dumb" timeouts, so I think we can leave it as is :) I know we have some examples in the codebase but considering there is many other cases too, let's leave it like it is :)